![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)



# Proyecto

### Equipo:

- Fran Zautzik
-


### Link de repositorio de GitHub: `https://github.com/bacchus00/laboratorio-mds`



In [1]:
import pandas as pd
import numpy as np


## 1. Introducción


A lo largo del siguiente informe se describe el proyecto de datos realizado para el Bodoque Bank. El proyecto tiene como objetivo la deteccion automatica de clientes fraudulentos en base a información relevante sobre ellos. El conjunto de datos que se utilizara para esta primera etapa del proyecto corresponde a The Bank Account Fraud (BAF), dataset usado para la evaluacion de metodos de deteccion de fraude bancario.

Cada fila del dataset corresponde a la informacion de un cliente (identificado por su id) junto con informacion sobre ellos. La informacion esta separada en distintos archivos los cuales seran unidos en un solo dataframe y no todos tienen las mismas columnas.

## 2. Lectura y Manejo de Datos




### 2.1 Cargar datos en dataframes

Primero importamos los archivos de extension '.parquet' que tienen los datos de BAF con el program de para obtencion de paquetes GNU 'wget'.

In [2]:
!wget -P / https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202/-/raw/main/df_1.parquet
!wget -P / https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202/-/raw/main/df_2.parquet
!wget -P / https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202/-/raw/main/df_email_phone.parquet

--2023-06-01 23:30:59--  https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202/-/raw/main/df_1.parquet
Resolving gitlab.com (gitlab.com)... 172.65.251.78, 2606:4700:90:0:f22e:fbec:5bed:a9b9
Connecting to gitlab.com (gitlab.com)|172.65.251.78|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 35809038 (34M) [application/octet-stream]
Saving to: ‘df_1.parquet’

f_1.parquet         100%[===================>]  34.15M  43.8MB/s    in 0.8s    

2023-06-01 23:31:00 (43.8 MB/s) - ‘df_1.parquet’ saved [35809038/35809038]

--2023-06-01 23:31:01--  https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202/-/raw/main/df_2.parquet
Resolving gitlab.com (gitlab.com)... 172.65.251.78, 2606:4700:90:0:f22e:fbec:5bed:a9b9
Connecting to gitlab.com (gitlab.com)|172.65.251.78|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 65248602 (62M) [application/octet-stream]
Saving to: ‘df_2.parquet’

f_2.parquet         100%[===================>]  62.23M  93.8MB/s 

Luego leemos los archivos importados para guardarlos en forma de dataframes de pandas.

In [3]:
df_1 = pd.read_parquet('df_1.parquet', engine='pyarrow')
df_2 = pd.read_parquet('df_2.parquet', engine='pyarrow')
df_email_phone = pd.read_parquet('df_email_phone.parquet', engine='pyarrow')

### 2.2 Exploración inicial

Inicialmente se explora superficialmente los 3 dataframes haciendo uso de la funcion auxiliar 'describe_df'.

In [4]:
def describe_df(df):
  print('Primeras 5 filas:')
  display(df.head())
  print('\nInformacion del contenido:')
  print(df.info())
  print(f'\nTiene {df.shape[0]} filas y {df.shape[1]} columnas.')

In [5]:
describe_df(df_1)

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]


Informacion del contenido:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 389782 entries, 104446 to 193209
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            389782 non-null  int64  
 1   fraud_bool                    389782 non-null  int64  
 2   income                        389782 non-null  float64
 3   prev_address_months_count     389782 non-null  int64  
 4   current_address_months_count  389782 non-null  int64  
 5   customer_age                  389782 non-null  int64  
 6   days_since_request            389782 non-null  float64
 7   intended_balcon_amount        389782 non-null  float64
 8   payment_type                  389782 non-null  object 
 9   zip_count_4w                  389782 non-null  int64  
 10  velocity_6h                   389782 non-null  float64
 11  velocity_24h                  389782 non-null  float64
 12  velocity_4w

In [6]:
describe_df(df_2)

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]


Informacion del contenido:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 747410 entries, 699954 to 384611
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            747410 non-null  int64  
 1   fraud_bool                    747410 non-null  int64  
 2   income                        747410 non-null  float64
 3   prev_address_months_count     747410 non-null  int64  
 4   current_address_months_count  747410 non-null  int64  
 5   customer_age                  747410 non-null  int64  
 6   days_since_request            747410 non-null  float64
 7   intended_balcon_amount        747410 non-null  float64
 8   payment_type                  747410 non-null  object 
 9   zip_count_4w                  747410 non-null  int64  
 10  velocity_6h                   747410 non-null  float64
 11  velocity_24h                  747410 non-null  float64
 12  velocity_4w

In [7]:
describe_df(df_email_phone)

Primeras 5 filas:


id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1


Informacion del contenido:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 934730 entries, 987231 to 986114
Data columns (total 7 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                934730 non-null  int64  
 1   name_email_similarity             934730 non-null  float64
 2   date_of_birth_distinct_emails_4w  934730 non-null  int64  
 3   email_is_free                     934730 non-null  int64  
 4   device_distinct_emails_8w         934730 non-null  int64  
 5   phone_home_valid                  934730 non-null  int64  
 6   phone_mobile_valid                934730 non-null  int64  
dtypes: float64(1), int64(6)
memory usage: 57.1 MB
None

Tiene 934730 filas y 7 columnas.


El primer dataframe 'df_1' tiene 389,782 filas y 29 columnas, no contiene valores nulos y consta principalmente de atributos de tipo numerico (int64 o float64), salvo 5 de tipo texto (object).

El segundo dataframe 'df_2' cuenta con 747,410 filas y 29 columnas, tampoco tiene valores nulos y tiene exactamente las mismas columnas que 'df_1'.

El tercer y ultimo dataframe, 'df_email_phone' 934,730 filas y 7 columnas, tampoco tiene valores faltantes y consta solo de valores numericos. Esta tabla solo comparte la columna 'id' con las dos tablas anteriores.

### 2.3 Unir dataframes

En seguida procedemos a simplemente concatenar las tablas 'df_1' y 'df_2' puesto que tienen exactamente la misma estructura de columnas.

In [8]:
df = pd.concat([df_1, df_2])
df.shape

(1137192, 29)

### 2.4 Unir nuevas variables

Despues, considerando que 'df_email_phone' tiene un esquema diferente al de las dos tablas anteriores, se hace un join de la tabla resultante anterior 'df' y la tabla 'df_email_phone' sobre la columna 'id'. El join anterior se hace de manera 'outer' para asi conservar todos los registros de ambas tablas.

In [9]:
df = df.merge(df_email_phone, how='outer', on='id')
df.shape

(1198151, 35)

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



Ahora verificamos la presencia de valores faltantes en las distintas columnas.

In [10]:
print('---number of NaNs per column:')
print(df.isna().sum())
print('\n---number of NaNs where name_email_similarity is also NaN')
print(df[df['name_email_similarity'].isna()].isna().sum())
#df[df.isna().any(axis=1)]

---number of NaNs per column:
id                                      0
fraud_bool                          60959
income                              60959
prev_address_months_count           60959
current_address_months_count        60959
customer_age                        60959
days_since_request                  60959
intended_balcon_amount              60959
payment_type                        60959
zip_count_4w                        60959
velocity_6h                         60959
velocity_24h                        60959
velocity_4w                         60959
bank_branch_count_8w                60959
employment_status                   60959
credit_risk_score                   60959
housing_status                      60959
bank_months_count                   60959
has_other_cards                     60959
proposed_credit_limit               60959
foreign_request                     60959
source                              60959
session_length_in_minutes           60959
devi

Observando el primer conteos es posible darse cuenta que 'id' no tiene valores faltantes. Adicionalmente se puede ver que todas las columnas de las primeras dos tablas tienen las misma cantidad de valores faltantes y todas las columnas de la tercera tabla tienen la misma cantidad de valores faltantes, esto tiene sentido considerando que en sus tablas originales no habia presencia de nulos.

Viendo ya el segundo conteo, donde se cuentan los caso nulos donde uno de los atributo de la tercera tabla es nulo (osea todos los demas atributos en esa tabla tambien), es posible ver que arroja el la misma cantidad para estos atributos pero para los demas antributos no hay valores faltantes. Lo anterior indica que enverdad las filas de valores nulos para atributos de la primera tabla son disjuntos de las filas de valores nulos para atributos de la segunda tabla, o enverdad que los casos nulos se deben puramente al outer join echo anteriormente, cosa que podria ser evitada con un inner join (a costa de perder cerca de 140,000 filas, 12% de las filas aproximadamente).

A continuacion verificamos el numero de filas duplicadas.

In [11]:
print(df.duplicated().sum()) #filas duplicadas segun todos su atributos
print(df.duplicated(subset='id').sum()) #filas duplicadas segun id

202334
202334


Debido a que los numeros anteriores son iguales las filas con los id iguales contienen exactamente la misma informacion en sus otros atributos.

### 2.6 Limpiar

Ahora eliminamos todos las filas con valores faltantes perdiendo un total de 140686 filas.

In [12]:
nrow_0 = df.shape[0]
df = df.dropna()
nrow_0 - df.shape[0]

140686

Luego eliminamos todas las filas duplicadas perdiendo otras 183694 filas.

In [13]:
nrow_0 = df.shape[0]
df = df.drop_duplicates()
nrow_0 - df.shape[0]

183694

In [14]:
describe_df(df)

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
0  729517         0.0     0.7                       -1.0   
2  149585         0.0     0.8                       -1.0   
4   64486         0.0     0.9                       -1.0   
5  825283         0.0     0.5                       -1.0   
7    8308         0.0     0.9                       -1.0   

   current_address_months_count  customer_age  days_since_request  \
0                         305.0          60.0            0.030059   
2                         140.0          50.0            0.015659   
4                         171.0          50.0            0.001409   
5                          85.0          30.0            0.027292   
7                          39.0          30.0            0.010945   

   intended_balcon_amount payment_type  zip_count_4w  ...  device_fraud_count  \
0               -1.599455           AC         990.0  ...                 0.0   
2                3.951994           AA        1269.0  ...                 0.0   
4               28.159779           AB        4430.0  ...                 0.0   
5               -1.310498           AB        1698.0  ...                 0.0   
7               -1.450972           AC         569.0  ...                 0.0   

   month        x1        x2 name_email_similarity  \
0    2.0 -0.245425  0.568811              0.883485   
2    7.0  0.009336 -2.096682              0.113208   
4    1.0  2.229616 -0.005823              0.792797   
5    3.0 -0.193945  0.861207              0.865082   
7    3.0 -2.440650  0.354986              0.450024   

   date_of_birth_distinct_emails_4w email_is_free  device_distinct_emails_8w  \
0                               2.0           1.0                        1.0   
2                               4.0           1.0                        1.0   
4                               4.0           0.0                        1.0   
5                              10.0           0.0                        1.0   
7                              13.0           1.0                        1.0   

   phone_home_valid  phone_mobile_valid  
0               0.0                 1.0  
2               1.0                 1.0  
4               0.0                 1.0  
5               0.0                 1.0  
7               0.0                 0.0  

[5 rows x 35 columns]


Informacion del contenido:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 873771 entries, 0 to 1137191
Data columns (total 35 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                873771 non-null  int64  
 1   fraud_bool                        873771 non-null  float64
 2   income                            873771 non-null  float64
 3   prev_address_months_count         873771 non-null  float64
 4   current_address_months_count      873771 non-null  float64
 5   customer_age                      873771 non-null  float64
 6   days_since_request                873771 non-null  float64
 7   intended_balcon_amount            873771 non-null  float64
 8   payment_type                      873771 non-null  object 
 9   zip_count_4w                      873771 non-null  float64
 10  velocity_6h                       873771 non-null  float64
 11  velocity_24h           

Y luego de esto quedamos con una tabla sin valores faltantes, ni duplicados de un tamaño de aproximadamente el 73% de el numero de ejemplos originales.

### 2.7 Segmentación Etaria

Agregamos la nueva columnas 'segmentacion_etaria' la cual simplifica la edad de los clientes a una variable categorica de 4 niveles.

In [15]:
def segmentacion_etaria_mapping(x):
  if x < 18: 
    return 'Joven'
  if x in range(18, 27):
    return 'Adulto-Joven'
  if x in range(27, 60):
    return 'Adulto'
  if x >= 60: 
    return 'Persona Mayor'

df['segmentacion_etaria'] = df['customer_age'].apply(segmentacion_etaria_mapping)
df['segmentacion_etaria'].value_counts()

Adulto           627027
Adulto-Joven     128374
Persona Mayor    107027
Joven             11343
Name: segmentacion_etaria, dtype: int64

## 3. Análisis Exploratorio de Datos

A continuacion se realiza una exploracion de datos con ayuda de la libreria 'ydata-profiling'

In [16]:
!pip install ydata_profiling

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 352.3/352.3 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.7/102.7 kB 13.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.5/679.5 kB 45.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 kB 29.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 455.4/455.4 kB 45.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 86.6 MB/s eta 0:00:00
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27081 sha256=7d93914f4e80dd200762271815e2b59fa91c53f5e74c48215cedb6bc753e64ec
  Stored in directory: /root/.cache/pip/wheels/dd/91/29/a79cecb328d01739e64017b6fb9a1ab9d8cb1853098ec5966d
Successfully built htmlmin
  Attempting uninstall: wordcloud
    Found existing installation: wordcloud 1.8.2.2
    

In [17]:
from ydata_profiling import ProfileReport

profile = ProfileReport(df, title="EDA")
profile.to_file("eda.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### 3.1 Análisis del EDA

En esta subseccion se hace un analisis de las variables del dataset, tanto de manera univariada como bivariada.

La variable 'id' consta de puros valores unicos, categoricos y esta distribuido de manera uniforme

La variable 'fraud_bool' un variable categorica de 2 niveles. Puede observarse que la clase esta altamente desbalanceada con teniendo un 91% de sus valores de la clase 0.

La variable 'foreign_request' es una variable categorica de dos niveles que esta altamente desbalanceada con cerca de un 84%	porciento de sus valores en 0.

La variable 'source' es una variable categorica de dos niveles que esta altamente desbalanceada con cerca de un 94%	porciento de sus valores en 'INTERNET.

La variable 'device_distinct_emails_8w' es una variable categorica de tres niveles que esta altamente desbalanceada con cerca de un 87%	porciento de sus valores en 1.

'month' tiene alrededor de 8% de sus valores en 0.

'bank_branch_count_8w' tiene cerca de 14% de sus valores en 0.

'x1' y 'x2' estan compuestos por valores unicos.

Observando la matriz de correlacion es posible notar que 'velocity_4w' esta negativa y fuertemente correlacionado con 'month' y 'segmentacion_etaria' esta fuertemente y positivamente correlacionado con 'customer_age' (cosa razonable considerando que una es una simplificacion de otra). 'prev_address_months_count' esta correlacionados negativamente 'current_address_months_count'. 'velocity_24h' esta positivamente correlacionado con 'velocity_4w' y negativamente correlacionado con 'month'. 'month' esta negativamente correlacionado con 'velocity_24h'. 'bank_branch_count_8w' esta correlacionado positivamente con 'bank_months_count'.
'credit_risk_score' esta correlacionado positivamente con 'proposed_credit_limit'.


## 4. Preprocesamiento

### 4.1 Declarar `ColumnTransformer`

A continuacion definimos los conjuntos de tipos de datos. Dentro de las variables categoricas hay varias variables binarias  que ya vienen codificadas numericamente, la variable 'source' igual es binaria por lo que podria ser util hacer un reemplazo a 0 y 1. Para las demas variables categoricas no hay un orden que se pueda intuir entre sus niveles, por lo que son variables categoricas solo nominales, no ordinales. ES por lo anterior que en estas variables se utilizara 'OneHotEncoding'

In [19]:
df['source'] = df['source'].replace(to_replace={'INTERNET':0, 'TELEAPP':1}).astype(int)

binary_vars = [
    'email_is_free',
    'phone_home_valid',
    'phone_mobile_valid',
    'has_other_cards',
    'foreign_request',
    'keep_alive_session',
    'source'
]

nominal_vars = list(df.select_dtypes(include='object').columns)
numerical_vars = [x for x in list(df.columns) if x not in (binary_vars+nominal_vars+['id', 'fraud_bool'])]

In [23]:
print(binary_vars)
print(nominal_vars)
print(numerical_vars)

['email_is_free', 'phone_home_valid', 'phone_mobile_valid', 'has_other_cards', 'foreign_request', 'keep_alive_session', 'source']
['payment_type', 'employment_status', 'housing_status', 'device_os', 'segmentacion_etaria']
['income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'credit_risk_score', 'bank_months_count', 'proposed_credit_limit', 'session_length_in_minutes', 'device_fraud_count', 'month', 'x1', 'x2', 'name_email_similarity', 'date_of_birth_distinct_emails_4w', 'device_distinct_emails_8w']


Luego, haciendo uso de ColumnTransformer definimos las transformaciones a realizar sobre las varibales nominales y dejamos pasar el resto de las variables intactas con 'passthrough'.

In [24]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

t = [("categoricalOHE", OneHotEncoder(sparse_output=False), nominal_vars)]
transformers = ColumnTransformer(transformers=t, remainder='passthrough')
transformers.set_output(transform='pandas')

ColumnTransformer(remainder='passthrough',
                  transformers=[('categoricalOHE',
                                 OneHotEncoder(sparse_output=False),
                                 ['payment_type', 'employment_status',
                                  'housing_status', 'device_os',
                                  'segmentacion_etaria'])])

### 4.2 Transformar datos

Ahora se aplica las transformaciones a realizar sobre el dataframe.

In [25]:
transformed_df = transformers.fit_transform(df)
transformed_df.columns

Index(['categoricalOHE__payment_type_AA', 'categoricalOHE__payment_type_AB',
       'categoricalOHE__payment_type_AC', 'categoricalOHE__payment_type_AD',
       'categoricalOHE__payment_type_AE',
       'categoricalOHE__employment_status_CA',
       'categoricalOHE__employment_status_CB',
       'categoricalOHE__employment_status_CC',
       'categoricalOHE__employment_status_CD',
       'categoricalOHE__employment_status_CE',
       'categoricalOHE__employment_status_CF',
       'categoricalOHE__employment_status_CG',
       'categoricalOHE__housing_status_BA',
       'categoricalOHE__housing_status_BB',
       'categoricalOHE__housing_status_BC',
       'categoricalOHE__housing_status_BD',
       'categoricalOHE__housing_status_BE',
       'categoricalOHE__housing_status_BF',
       'categoricalOHE__housing_status_BG', 'categoricalOHE__device_os_linux',
       'categoricalOHE__device_os_macintosh',
       'categoricalOHE__device_os_other', 'categoricalOHE__device_os_windows',
       

## 5. Visualización en Baja Dimensionalidad

### 5.1 Muestrear dataframe

Enseguida se selecciona una muestra aleatoria de 25000 filas y se separa las variables explicativas (X) de la variable target (y)

In [68]:
sample_25k = transformed_df.sample(25000).reset_index(drop=True)

X = sample_25k.loc[:, sample_25k.columns!='remainder__fraud_bool']
y = sample_25k['remainder__fraud_bool']

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

Recuperamos los indices de las distintas categorias de las segmentacion etaria y extraemos sus variables correpondientes.

In [69]:
i_j = list(X[X['categoricalOHE__segmentacion_etaria_Joven']==1].index)
i_aj = list(X[X['categoricalOHE__segmentacion_etaria_Adulto-Joven']==1].index)
i_a = list(X[X['categoricalOHE__segmentacion_etaria_Adulto']==1].index)
i_pm = list(X[X['categoricalOHE__segmentacion_etaria_Persona Mayor']==1].index)

X_j = X.loc[i_j]
X_aj = X.loc[i_aj]
X_a = X.loc[i_a]
X_pm = X.loc[i_pm]

Luego se hace uso del metodo PCA para reducir la dimensionalidad de los datos a sus dos componentes principales para cada uno de los rangos etarios.

In [70]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
pca.fit(X_j)
X_pca_j = pca.transform(X_j)
X_pca_j = pd.DataFrame(X_pca_j, columns = ['Component1','Component2'])
X_pca_j

pca = PCA(n_components=2)
pca.fit(X_aj)
X_pca_aj = pca.transform(X_aj)
X_pca_aj = pd.DataFrame(X_pca_aj, columns = ['Component1','Component2'])
X_pca_aj

pca = PCA(n_components=2)
pca.fit(X_a)
X_pca_a = pca.transform(X_a)
X_pca_a = pd.DataFrame(X_pca_a, columns = ['Component1','Component2'])
X_pca_a

pca = PCA(n_components=2)
pca.fit(X_pm)
X_pca_pm = pca.transform(X_pm)
X_pca_pm = pd.DataFrame(X_pca_pm, columns = ['Component1','Component2'])
X_pca_pm

Component1   Component2
0    -424672.639445 -1476.997261
1    -391454.102610  3413.470046
2     383147.727479  3976.891333
3     -84956.782738 -1072.975152
4    -333554.708488 -1949.173024
...             ...          ...
3018 -271574.678688  -559.011182
3019  438307.244306 -1776.136635
3020   20873.701885  4568.961465
3021 -220674.727531 -2584.524093
3022 -291878.806730 -3782.962539

[3023 rows x 2 columns]

### 5.3 Visualizar según rangos etarios

Una vez que tenemos los datos proyectados a dos dimensiones realizamos scatterplots en las dos componentes principales usando el color para indicar la categoria fraud_bool del dato.

In [71]:
import plotly.express as px

px.scatter(X_pca_j, x='Component1', y='Component2', color=y.loc[i_j])

In [72]:
px.scatter(X_pca_aj, x='Component1', y='Component2', color=y.loc[i_aj])

In [73]:
px.scatter(X_pca_a, x='Component1', y='Component2', color=y.loc[i_a])

In [74]:
px.scatter(X_pca_pm, x='Component1', y='Component2', color=y.loc[i_pm])

Lamentablemente con la reduccion de dimensionalidad no se logra observar agrupaciones claramente separables por algun criterio

## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

A continuacion se ocupa el metodo de deteccion de anomalias 'EllipticEnvelope' para ver si es posible detectar los datos correspondientes a clientes fraudulentos mediante este metodo. Los resultados entregados se transforman para que 1 corresponda a un outlier y 0 a un inlier

In [75]:
from sklearn.cluster import KMeans
from sklearn.covariance import EllipticEnvelope
cov_j = EllipticEnvelope().fit(X_j)

/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:745: UserWarning:

The covariance matrix associated to your dataset is not full rank



In [76]:
cov_aj = EllipticEnvelope().fit(X_aj)

/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:745: UserWarning:

The covariance matrix associated to your dataset is not full rank



In [77]:
cov_a = EllipticEnvelope().fit(X_a)

/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:745: UserWarning:

The covariance matrix associated to your dataset is not full rank



In [78]:
cov_pm = EllipticEnvelope().fit(X_pm)

/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:745: UserWarning:

The covariance matrix associated to your dataset is not full rank



### 6.2 Agregar resultados a dataframe de muestreo

In [91]:
pca = PCA(n_components=2)
pca.fit(X_j)
X_pca_j = pca.transform(X_j)
X_pca_j = pd.DataFrame(X_pca_j, columns = ['Component1','Component2'])
X_pca_j['outlier'] = (cov_j.predict(X_j)*(-1)+1)/2

In [92]:
pca = PCA(n_components=2)
pca.fit(X_aj)
X_pca_aj = pca.transform(X_aj)
X_pca_aj = pd.DataFrame(X_pca_aj, columns = ['Component1','Component2'])
X_pca_aj['outlier'] = (cov_j.predict(X_aj)*(-1)+1)/2

In [93]:
pca = PCA(n_components=2)
pca.fit(X_a)
X_pca_a = pca.transform(X_a)
X_pca_a = pd.DataFrame(X_pca_a, columns = ['Component1','Component2'])
X_pca_a['outlier'] = (cov_j.predict(X_a)*(-1)+1)/2

In [94]:
pca = PCA(n_components=2)
pca.fit(X_pm)
X_pca_pm = pca.transform(X_pm)
X_pca_pm = pd.DataFrame(X_pca_pm, columns = ['Component1','Component2'])
X_pca_pm['outlier'] = (cov_j.predict(X_pm)*(-1)+1)/2

### 6.3 Visualizar según rangos etarios

In [95]:
px.scatter(X_pca_j, x='Component1', y='Component2', color='outlier')

In [96]:
px.scatter(X_pca_aj, x='Component1', y='Component2', color='outlier')

In [97]:
px.scatter(X_pca_a, x='Component1', y='Component2', color='outlier').

In [98]:
px.scatter(X_pca_pm, x='Component1', y='Component2', color='outlier')

### 6.4 Calcular ratios y responder

A continuacion calculamos el precision y el recall del modelo de deteccion de anomalias.

In [126]:
tp = X_pca_j['outlier'].add(y.loc[i_j].reset_index(drop=True)).value_counts().to_dict()[2]+X_pca_aj['outlier'].add(y.loc[i_aj].reset_index(drop=True)).value_counts().to_dict()[2]+X_pca_a['outlier'].add(y.loc[i_a].reset_index(drop=True)).value_counts().to_dict()[2]+X_pca_pm['outlier'].add(y.loc[i_pm].reset_index(drop=True)).value_counts().to_dict()[2]
pp = X_pca_j['outlier'].sum() + X_pca_aj['outlier'].sum() + X_pca_a['outlier'].sum() + X_pca_pm['outlier'].sum()
P = tp/fd
P

0.014133760746029433

In [127]:
p = y.loc[i_j].sum() + y.loc[i_aj].sum() + y.loc[i_a].sum() + y.loc[i_pm].sum()
R = tp/p
R

0.35144927536231885

Se observa que los valores son excesivamente bajos y que porlotanto nuestro modelos, porlomenos con la data disṕnible y con el preprocesamiento escogido no es capaz de predecir satisfactoriamente clientes faudulentos.